src/main/java/macrobase/analysis/contextualoutlier/Context.java

```diff
@@ -108,78 +181,109 @@ public boolean containDatum(Datum datum){
 	 * Join this Context with other Context, can only be joined if the first (k-1) dimensions
 	 * have the same interval, and the last dimension has different interval
 	 * 
-	 * If not, return null
+	 * can be joined only if the size of the context is at least minSize
 	 * 
 	 * 
 	 * @param other
 	 * @return
 	 */
-	public Context join(Context other){
+	public Context join(Context other, List<Datum> data, double tau){
 		
-		SortedMap<Integer,Interval> newDimension2Interval = new TreeMap<Integer,Interval>();
+		List<Integer> newDimensions = new ArrayList<Integer>();
+		List<Interval> newIntervals = new ArrayList<Interval>();
 		
 		
-		List<Integer> dimensions1 = getDimensions();
-		List<Integer> dimensions2 = other.getDimensions();
+		List<Integer> dimensions1 = dimensions;
+		List<Integer> dimensions2 = other.dimensions;
 		if(dimensions1.size() != dimensions2.size())
 			return null;
 		
 		for(int i = 0; i < dimensions1.size(); i++){
 			int dimension1 = dimensions1.get(i);
 			int dimension2 = dimensions2.get(i);
-			Interval interval1 = dimension2Interval.get(dimension1);
-			Interval interval2 = other.dimension2Interval.get(dimension2);
+			Interval interval1 = intervals.get(i);
+			Interval interval2 = other.intervals.get(i);
 		
 			if(i !=dimensions1.size() - 1 ){
 				if(dimension1 != dimension2)
 					return null;
 					if(interval1 != interval2)
 					return null;
-				newDimension2Interval.put(dimension1, interval1);
+					newDimensions.add(dimension1);
+					newIntervals.add(interval1);
+				
 			}else{
-					newDimension2Interval.put(dimension1, interval1);
-				newDimension2Interval.put(dimension2, interval2);
+				newDimensions.add(dimension1);
+				newIntervals.add(interval1);
+				
+				newDimensions.add(dimension2);
+				newIntervals.add(interval2);
+				
 			}
 			
 		}
 	
-		Context newUnit = new Context(newDimension2Interval);
-		
-		//merge two sorted tids
-		int index1 = 0;
-		int index2 = 0;
-		while(index1 < tids.size() && index2 < other.tids.size()){
-			int tid1 = tids.get(index1);
-			int tid2 = other.tids.get(index2);
-			if(tid1 == tid2){
-				newUnit.addTID(tid1);
-				index1++;
-				index2++;
-			}else if(tid1 < tid2){
-				index1++;
-			}else{
-				index2++;
-			}
+		Context newUnit = new Context(newDimensions, newIntervals, this, other);
+		
+		
+		if(ContextPruning.densityPruning(newUnit, tau)){
+			return null;
+		}
+		
+		if(ContextPruning.dependencyPruning(newUnit)){
+			return null;
 		}
 		
 		return newUnit;
 	}
 	
+
 	public String print(DatumEncoder encoder){
```

Comment:
I have a toString() method in Context.java. The print() function is to print the context using original data values, instead of using the encoded integer values as toString() does.